File: app/local_plugins/agent_runtime/ios/Classes/AgentRuntimePlugin.swift
import Flutter
import UIKit
import AVFoundation

/// AgentRuntimePlugin — iOS Flutter Plugin 入口
///
/// 命令通道（Flutter→Native）：MethodChannel "agent_runtime/commands"
/// 事件通道（Native→Flutter）：EventChannel  "agent_runtime/events"
public class AgentRuntimePlugin: NSObject, FlutterPlugin, AgentEventSink {

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSinkStream: FlutterEventSink?

    private let manager = AgentRuntimeManager.shared

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AgentRuntimePlugin()

        instance.methodChannel = FlutterMethodChannel(
            name: "agent_runtime/commands",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        instance.eventChannel = FlutterEventChannel(
            name: "agent_runtime/events",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel?.setStreamHandler(instance)

        instance.manager.eventSink = instance
    }

    // ─────────────────────────────────────────────────
    // FlutterMethodCallDelegate
    // ─────────────────────────────────────────────────

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "startSession":
            let config = args.toAgentSessionConfig()
            Task { await manager.startSession(config: config) }
            result(nil)
        case "stopSession":
            let sessionId = args["sessionId"] as! String
            Task { await manager.stopSession(sessionId: sessionId) }
            result(nil)
        case "sendText":
            Task {
                await manager.sendText(
                    sessionId: args["sessionId"] as! String,
                    requestId: args["requestId"] as! String,
                    text: args["text"] as! String
                )
            }
            result(nil)
        case "interrupt":
            Task { await manager.interrupt(sessionId: args["sessionId"] as! String) }
            result(nil)
        case "setInputMode":
            Task {
                await manager.setInputMode(
                    sessionId: args["sessionId"] as! String,
                    mode: args["mode"] as! String
                )
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ─────────────────────────────────────────────────
    // AgentEventSink 实现
    // ─────────────────────────────────────────────────

    func onSttEvent(_ event: SttEventData) {
        pushEvent([
            "type": "stt", "sessionId": event.sessionId, "requestId": event.requestId,
            "kind": event.kind, "text": event.text as Any,
            "errorCode": event.errorCode as Any, "errorMessage": event.errorMessage as Any,
        ])
    }

    func onLlmEvent(_ event: LlmEventData) {
        pushEvent([
            "type": "llm", "sessionId": event.sessionId, "requestId": event.requestId,
            "kind": event.kind,
            "textDelta": event.textDelta as Any, "thinkingDelta": event.thinkingDelta as Any,
            "toolCallId": event.toolCallId as Any, "toolName": event.toolName as Any,
            "toolArgumentsDelta": event.toolArgumentsDelta as Any,
            "toolResult": event.toolResult as Any, "fullText": event.fullText as Any,
            "errorCode": event.errorCode as Any, "errorMessage": event.errorMessage as Any,
        ])
    }

    func onTtsEvent(_ event: TtsEventData) {
        pushEvent([
            "type": "tts", "sessionId": event.sessionId, "requestId": event.requestId,
            "kind": event.kind,
            "progressMs": event.progressMs as Any, "durationMs": event.durationMs as Any,
            "errorCode": event.errorCode as Any, "errorMessage": event.errorMessage as Any,
        ])
    }

    func onStateChanged(sessionId: String, state: String, requestId: String?) {
        pushEvent([
            "type": "stateChanged", "sessionId": sessionId,
            "state": state, "requestId": requestId as Any,
        ])
    }

    func onError(sessionId: String, errorCode: String, message: String, requestId: String?) {
        pushEvent([
            "type": "error", "sessionId": sessionId,
            "errorCode": errorCode, "message": message, "requestId": requestId as Any,
        ])
    }

    private func pushEvent(_ data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSinkStream?(data)
        }
    }
}

// ─────────────────────────────────────────────────
// FlutterStreamHandler（EventChannel）
// ─────────────────────────────────────────────────

extension AgentRuntimePlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSinkStream = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSinkStream = nil
        return nil
    }
}

// ─────────────────────────────────────────────────
// 扩展：[String: Any] → AgentSessionConfig
// ─────────────────────────────────────────────────

private extension Dictionary where Key == String, Value == Any {
    func toAgentSessionConfig() -> AgentSessionConfig {
        AgentSessionConfig(
            sessionId: self["sessionId"] as! String,
            agentId: self["agentId"] as! String,
            inputMode: self["inputMode"] as! String,
            sttPluginName: self["sttPluginName"] as! String,
            ttsPluginName: self["ttsPluginName"] as! String,
            llmPluginName: self["llmPluginName"] as! String,
            stsPluginName: self["stsPluginName"] as? String,
            sttConfigJson: self["sttConfigJson"] as! String,
            ttsConfigJson: self["ttsConfigJson"] as! String,
            llmConfigJson: self["llmConfigJson"] as! String,
            stsConfigJson: self["stsConfigJson"] as? String
        )
    }
}

File: app/local_plugins/agent_runtime/android/src/main/kotlin/com/aiagent/agent_runtime/AgentRuntimePlugin.kt
package com.aiagent.agent_runtime

import android.content.*
import android.os.IBinder
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*
import org.json.JSONObject

/**
 * AgentRuntimePlugin — Flutter Plugin 入口
 *
 * 命令通道（Flutter→Native）：接收 startSession / stopSession / sendText / interrupt / setInputMode
 * 事件通道（Native→Flutter）：推送 STT/LLM/TTS 事件和状态变更
 *
 * Pigeon 生成代码就绪后，MethodChannel 手动分发可替换为 Pigeon 自动绑定。
 */
class AgentRuntimePlugin : FlutterPlugin, AgentEventSink {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSinkStream: EventChannel.EventSink? = null

    private var service: AgentRuntimeService? = null
    private var isBound = false
    private lateinit var context: Context

    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // ─────────────────────────────────────────────────
    // FlutterPlugin 生命周期
    // ─────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, "agent_runtime/commands")
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "startSession" -> {
                    val config = call.arguments<Map<*, *>>()!!.toAgentSessionConfig()
                    startServiceIfNeeded()
                    service?.startSession(config)
                    result.success(null)
                }
                "stopSession" -> {
                    service?.stopSession(call.argument("sessionId")!!)
                    result.success(null)
                }
                "sendText" -> {
                    service?.sendText(
                        call.argument("sessionId")!!,
                        call.argument("requestId")!!,
                        call.argument("text")!!,
                    )
                    result.success(null)
                }
                "interrupt" -> {
                    service?.interrupt(call.argument("sessionId")!!)
                    result.success(null)
                }
                "setInputMode" -> {
                    service?.setInputMode(
                        call.argument("sessionId")!!,
                        call.argument("mode")!!,
                    )
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }

        eventChannel = EventChannel(binding.binaryMessenger, "agent_runtime/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSinkStream = events
            }
            override fun onCancel(arguments: Any?) {
                eventSinkStream = null
            }
        })

        bindService()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        if (isBound) context.unbindService(serviceConnection)
        mainScope.cancel()
    }

    // ─────────────────────────────────────────────────
    // AgentEventSink 实现（Native→Flutter 事件推送）
    // ─────────────────────────────────────────────────

    override fun onSttEvent(event: SttEventData) {
        pushEvent(mapOf(
            "type" to "stt",
            "sessionId" to event.sessionId,
            "requestId" to event.requestId,
            "kind" to event.kind,
            "text" to event.text,
            "errorCode" to event.errorCode,
            "errorMessage" to event.errorMessage,
        ))
    }

    override fun onLlmEvent(event: LlmEventData) {
        pushEvent(mapOf(
            "type" to "llm",
            "sessionId" to event.sessionId,
            "requestId" to event.requestId,
            "kind" to event.kind,
            "textDelta" to event.textDelta,
            "thinkingDelta" to event.thinkingDelta,
            "toolCallId" to event.toolCallId,
            "toolName" to event.toolName,
            "toolArgumentsDelta" to event.toolArgumentsDelta,
            "toolResult" to event.toolResult,
            "fullText" to event.fullText,
            "errorCode" to event.errorCode,
            "errorMessage" to event.errorMessage,
        ))
    }

    override fun onTtsEvent(event: TtsEventData) {
        pushEvent(mapOf(
            "type" to "tts",
            "sessionId" to event.sessionId,
            "requestId" to event.requestId,
            "kind" to event.kind,
            "progressMs" to event.progressMs,
            "durationMs" to event.durationMs,
            "errorCode" to event.errorCode,
            "errorMessage" to event.errorMessage,
        ))
    }

    override fun onStateChanged(sessionId: String, state: String, requestId: String?) {
        pushEvent(mapOf(
            "type" to "stateChanged",
            "sessionId" to sessionId,
            "state" to state,
            "requestId" to requestId,
        ))
    }

    override fun onError(sessionId: String, errorCode: String, message: String, requestId: String?) {
        pushEvent(mapOf(
            "type" to "error",
            "sessionId" to sessionId,
            "errorCode" to errorCode,
            "message" to message,
            "requestId" to requestId,
        ))
    }

    private fun pushEvent(data: Map<String, Any?>) {
        mainScope.launch {
            eventSinkStream?.success(data)
        }
    }

    // ─────────────────────────────────────────────────
    // Service 绑定
    // ─────────────────────────────────────────────────

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
            service = (binder as AgentRuntimeService.LocalBinder).getService()
            service?.eventSink = this@AgentRuntimePlugin
            isBound = true
        }
        override fun onServiceDisconnected(name: ComponentName?) {
            service = null
            isBound = false
        }
    }

    private fun startServiceIfNeeded() {
        val intent = Intent(context, AgentRuntimeService::class.java)
        context.startForegroundService(intent)
        if (!isBound) bindService()
    }

    private fun bindService() {
        val intent = Intent(context, AgentRuntimeService::class.java)
        context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
    }
}

// ── 扩展函数：Map → AgentSessionConfig ────────────────────────────────────

private fun Map<*, *>.toAgentSessionConfig() = AgentSessionConfig(
    sessionId = this["sessionId"] as String,
    agentId = this["agentId"] as String,
    inputMode = this["inputMode"] as String,
    sttPluginName = this["sttPluginName"] as String,
    ttsPluginName = this["ttsPluginName"] as String,
    llmPluginName = this["llmPluginName"] as String,
    stsPluginName = this["stsPluginName"] as? String,
    sttConfigJson = this["sttConfigJson"] as String,
    ttsConfigJson = this["ttsConfigJson"] as String,
    llmConfigJson = this["llmConfigJson"] as String,
    stsConfigJson = this["stsConfigJson"] as? String,
)
